text/0000-gen-fn.md

```diff
@@ -0,0 +1,429 @@
+[motivation]: #motivation
+
+The main motivation of this RFC is to reserve a new keyword in the 2024 edition.
+The feature used by the keyword described here should be treated as an e-RFC for
+experimentation on nightly. I would like to avoid discussion of the semantics
+provided here, deferring that discussion until during the experimental
+implementation work.
+
+Writing iterators manually can be very painful. Many iterators can be written by
+chaining `Iterator` methods, but some need to be written as a `struct` and have
+`Iterator` implemented for them. Some of the code that is written this way
+pushes people to avoid iterators and instead execute a `for` loop that eagerly
+writes values to mutable state. With this RFC, one can write the `for` loop
+and still get a lazy iterator of values.
+
+As an example, here are multiple ways to write an iterator over something that contains integers,
+only keep the odd integers, and multiply all of them by 2:
+
+```rust
+// `Iterator` methods
+fn odd_dup(values: impl Iterator<Item = u32>) -> impl Iterator<Item = u32> {
+    values.filter(|value| value.is_odd()).map(|value| value * 2)
+}
+
+// `struct` and manual `impl`
+fn odd_dup(values: impl Iterator<Item = u32>) -> impl Iterator<Item = u32> {
+    struct Foo<T>(T);
+    impl<T: Iterator<Item = u32>> Iterator<Item = u32> for Foo<T> {
+        type Item = u32;
+        fn next(&mut self) -> Option<u32> {
+            loop {
+                let value = self.0.next()?;
+                if value.is_odd() {
+                    return Some(x * 2)
+                }
+            }
+        }
+    }
+    Foo(values)
+}
+
+// `gen block`
+fn odd_dup(values: impl Iterator<Item = u32>) -> impl Iterator<Item = u32> {
+    #[rustc_gen] {
+        for value in values {
+            if value.is_odd() {
+                yield value * 2;
+            }
+        }
+    }
+}
+
+// `gen fn`
+#[rustc_gen]
+fn odd_dup(values: impl Iterator<Item = u32>) -> u32 {
+    for value in values {
+        if value.is_odd() {
+            yield value * 2;
+        }
+    }
+}
+```
+
+Iterators created with `#[rustc_gen]` return `None` once they `return` (implicitly at the end of the scope or explicitly with `return`).
+See [the unresolved questions][#unresolved-questions] for whether `#[rustc_gen]` iterators are fused or may behave strangely after having returned `None` once.
+Under no circumstances will it be undefined behavior if `next` is invoked again after having gotten a `None`.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## New keyword
+
+Starting in the 2024 edition, `gen` is a keyword that cannot be used for naming any items or bindings. This means during the migration to the 2024 edition, all variables, functions, modules, types, ... named `gen` must be renamed.
+
+## Returning/finishing an iterator
+
+`#[rustc_gen]` block's trailing expression must be of the unit type or the block must diverge before reaching its end.
```

Comment:
Is "a block's trailing expression must be of the unit type" sufficient to imply that uses of `return` within the block must either be given `()` or no expression at all?

I personally suspect not, since I can imagine linguistic constructs that would do some kind of coercion of the unit value from the trailing expression while still allowing `return` to take a non-unit expression as input.

And therefore, based on that suspicion, I recommend this bit of text be revised to explicitly say (somehow) that `return` within the block is restricted to unit. (E.g. " uses of `return` within the block must either be given `()` or no expression at all.")

Reply:
My proposal is to treat `return` and finishing the block exactly the same and require them both to be returning unit

Reply:
We've now revised this section to be more clear and comprehensive with respect to what is allowed in the trailing position and as an argument to return.  Thanks for pointing out the ambiguities here.

(I'm going to mark this conversation as resolved, but of course, if you have any observations about the new language, please open one for that.)